File: solutions/384. Shuffle an Array.cpp
class Solution {
private:
    std::vector<int> array;
    std::vector<int> arr_copy;
    
public:
    Solution(std::vector<int>& nums):
        array(std::move(nums)),
        arr_copy(std::vector<int> { array.begin(), array.end() }) {}
    
    /** Resets the array to its original configuration and return it. */
    std::vector<int> reset() {
        array = arr_copy;
        arr_copy = std::vector<int> { arr_copy.begin(), arr_copy.end() };
        return arr_copy;
    }
    
    /** Returns a random shuffling of the array. */
    std::vector<int> shuffle() {
        std::mt19937 rng { std::random_device{}() };
        std::shuffle(begin(array), end(array), rng);
        return array;
    }
};
​
/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(nums);
 * vector<int> param_1 = obj->reset();
 * vector<int> param_2 = obj->shuffle();
 */

File: ..solutions/384. Shuffle an Array.java
class Solution {
​
    private int[] originalArray;
    private int[] temp;
    private Random random;
    public Solution(int[] nums) {
        this.originalArray = nums.clone();
        this.temp = nums.clone();
        this.random = new Random();
    }
    
    /** Resets the array to its original configuration and return it. */
    public int[] reset() {
        this.temp = this.originalArray.clone(); // .clone() is the fucking hell. to use otherwize object will shallow copy.
        // this deeply copies
        return this.temp;
    }
    
    /** Returns a random shuffling of the array. */
    public int[] shuffle() {
        for (int i = this.temp.length - 1; i >= 0; i--) {
            int randomIndex = this.randomIndexGenerator(i+1);
            this.swap(i, randomIndex);
        }
        
        return this.temp;
    }
    
    private void swap(int i, int j) {
        int t = this.temp[i];
        this.temp[i] = this.temp[j];
        this.temp[j] = t;
    }
    
    private int randomIndexGenerator(int rangei) { return this.random.nextInt(rangei) ; }
}
​
/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * int[] param_1 = obj.reset();
 * int[] param_2 = obj.shuffle();
 */
